include/ignition/common/FlagSet.hh

```diff
@@ -0,0 +1,348 @@
+  {
+    FlagSet ret(*this);
+    ret.flags |= _val.flags;
+
+    return ret;
+  }
+
+  /// \brief Return a negation of this FlagSet.
+  /// \return The negated FlagSet.
+  public: FlagSet operator~() const
+  {
+    FlagSet cp(*this);
+    cp.flags.flip();
+
+    return cp;
+  }
+
+  /// \brief Return true if at least one flag is set to true.
+  /// \return Whether at least one flag is set.
+  public: explicit operator bool() const
+  {
+    return flags.any();
+  }
+
+  // Methods from std::bitset.
+
+  public: bool operator==(const FlagSet& _o) const
+  {
+    return flags == _o.flags;
+  }
+
+  public: bool operator!=(const FlagSet& _o) const
+  {
+    return !(*this == _o);
+  }
+
+  /// \brief Return the total number of flags represented by this FlagSet.
+  /// \note This corresponds to the value of the '_' element of the enum.
+  /// \return The number of flags in this FlagSet.
+  /// \sa Count()
+  public: std::size_t Size() const
+  {
+    return flags.size();
+  }
+
+  /// \brief Return the number of flags set to true.
+  /// \return The number of true flags in this FlagSet.
+  /// \sa Size()
+  public: std::size_t Count() const
+  {
+    return flags.count();
+  }
+
+  /// \brief Set all flags to true.
+  /// \return This.
+  public: FlagSet& Set()
+  {
+    flags.set();
+    return *this;
+  }
+
+  /// \brief Set all flags to false.
+  /// \return This.
+  public: FlagSet& Reset()
+  {
+    flags.reset();
+    return *this;
+  }
+
+  /// \brief Set all flags to their negation.
+  /// \return This.
+  public: FlagSet& Flip()
+  {
+    flags.flip();
+    return *this;
+  }
+
+  /// \brief Set the given flag to the specified value.
+  /// \param[in] _val The flag to set.
+  /// \param[in] _value The value to set.
+  /// \return This.
+  public: FlagSet& Set(const T& _val, bool _value = true)
+  {
+    flags.set(static_cast<UnderlyingType>(_val), _value);
+    return *this;
+  }
+
+  /// \brief Set the given flag to false.
+  /// \param[in] _val The flag to set.
+  /// \return This.
+  public: FlagSet& Reset(const T& _val)
+  {
+    flags.reset(static_cast<UnderlyingType>(_val));
+    return *this;
+  }
+
+  /// \brief Negate the given flag.
+  /// \param[in] _val The flag to negate.
+  /// \return This.
+  public: FlagSet& Flip(const T& _val)
+  {
+    flags.flip(static_cast<UnderlyingType>(_val));
+    return *this;
+  }
+
+  /// \brief Test whether any flag is set.
+  /// \return True if at least one flag is set to true.
+  public: bool Any() const
+  {
+    return flags.any();
+  }
+
+  /// \brief Test whether all flags are set.
+  /// \return True if at all flags are set to true.
+  public: bool All() const
+  {
+    return flags.all();
+  }
+
+  /// \brief Test whether no flag is set.
+  /// \return True if at no flag is set to true.
+  public: bool None() const
+  {
+    return flags.none();
+  }
+
+  /// \brief Retrurn a FlagSet with all flags set to true.
+  public: static FlagSet ALL()
```

Comment:
I think also that `All` vs `ALL` may be a bit confusing.

Style Nit: We haven't been using all-caps for static methods (that I'm aware of).

Reply:
No, static methods follow the same `CamelCase` rule as other member functions.

Reply:
Okay, what could be a better name if we'd want to reserve `All()` to the corresponding `all()` function of the bitset? Also `NONE` should be renamed. Some suggestions (I don't have a favorite one yet):

1. AllTrue + AllFalse
1. AllTrue + NoneTrue
1. AllEnabled + NoneEnabled
1. AllEnabled + AllDisabled
1. AllSet + NoneSet

but all of these names still don't clearly differ from the names usually used for getters. The idea was to provide something like `ignition::math::Vector3d::One` to avoid construction patterns like `FlagSet<>().Set()`.

The `NONE` helper is not really needed as the default behavior of the constructor is to set no flags, but I prefered parity of the offered functions.

Another option could be to add a `FlagSet<>(bool)` constructor which would take a parameter telling whether all flags should be set or not. But I think it's less readable and verbose.

Reply:
I like `All`, `Any`, `None` as our three checks, as it mirrors STL's bitset, so I think we're good there.

I think my preference for the `static` methods would be `AllEnabled`, `AllDisabled`, but I think it would be worth asking some non-native English speakers to see if it makes as much sense to them.

Reply:
I am non-native :-D But it's a good idea. Where to ask? I assume few people will now have others around to ask them on a coffee break...

Reply:
I'm pinging our company chat to recruit some eyeballs.

Reply:
Keeping `All` / `Any` / `Set` as parallels to the `std::bitset` API also make sense to me.

For the static methods, `AllSet` / `NoneSet` make slightly more sense to me - the class is called `FlagSet` afterall. But it's not a strong preference.

Reply:
Okay, renamed according to Louise.